src/backend.py

```diff
@@ -0,0 +1,726 @@
+import Pico_Led
+import faults
+
+#
+# Constants
+#
+rtc = RTC()
+ram_access = bytearray_at(SRAM_DATA_BASE,SRAM_DATA_LENGTH)
+WIFI_MAX_ATTEMPTS = 12
+AP_NAME = "Warped Pinball"
+# Authentication variables
+current_challenge = None
+challenge_timestamp = None
+CHALLENGE_EXPIRATION_SECONDS = 60
+
+
+#
+# Standardized Route Functions
+#
+def route_wrapper(func):
+    def wrapped_route(request):
+        gc_collect()
+        try:
+            response = func(request)
+
+            if response is None:
+                return "ok", 200
+            return response
+
+        except Exception as e:
+            msg = f"Error in {func.__name__}: {e}"
+            print(msg)
+            print(request)
+            print(request.query)
+            return msg, 500
+
+        finally:
+            gc_collect()
+    return wrapped_route
+
+def add_route(path, method="GET", auth=False, cool_down_seconds=0, single_instance=False):
+    '''Decorator to add a route to the server with gc_collect() and error handling'''
+    # If auth is True only allow a single instance of the route to run at a time
+    if auth:
+        single_instance = True
+
+    def decorator(func):
+        if cool_down_seconds > 0 or single_instance:
+            func = cool_down(cool_down_seconds, single_instance)(func)
+
+        if auth:
+            func = require_auth(func)
+
+        wrapped = route_wrapper(func)
+
+        @server.route(path, methods=[method])
+        def route(request):
+            return wrapped(request)
+        
+        return route
+    return decorator
+
+def get_content_type(file_path):
+    content_type_mapping = {
+        '.css': 'text/css',
+        '.js': 'application/javascript',
+        '.html': 'text/html',
+        '.png': 'image/png',
+        '.jpg': 'image/jpeg',
+        '.jpeg': 'image/jpeg',
+        '.gif': 'image/gif',
+        '.gz': 'application/gzip'
+    }
+    for extension, content_type in content_type_mapping.items():
+        if file_path.endswith(extension):
+            return content_type
+    return 'application/octet-stream'
+
+def create_file_handler(file_path):
+    def file_stream_generator():
+        gc_collect()
+        with open(file_path, 'rb') as f:
+            while True:
+                chunk = f.read(1024)  # Read in chunks of 1KB
+                if not chunk:
+                    break
+                yield chunk
+        gc_collect()
+
+    def file_handler(request):
+        headers = {
+            'Content-Type': get_content_type(file_path),
+            'Connection': 'close'
+        }
+        # TODO add 'Cache-Control': 'max-age=86400'  #to headers  in order to cache for a day
+        # not sure how we would clear this cache, we might want to consider options first
+        return file_stream_generator(), 200, headers
+
+    # Wrap the file_handler with the same error/memory logic
+    return route_wrapper(file_handler)
+
+def cool_down(cool_down_seconds=0, single_instance=False):
```

Comment:
This is how we handle the cool_down_seconds and single instnace arguments in `@add_route` calls. I've found single instance to be most useful but cool down might be important for future features